pycuda/gpuarray.py

```diff
@@ -1049,6 +1064,112 @@ class Info(Record):
 
 # }}}
 
+def _discontiguous_order(ary):
+    if isinstance(ary, GPUArray):
+        if ary.flags.c_contiguous: return 'C'
+        elif ary.flags.f_contiguous: return 'F'
+    elif isinstance(ary, np.ndarray):
+        if ary.flags['C_CONTIGUOUS']: return 'C'
+        elif ary.flags['F_CONTIGUOUS']: return 'F'
+
+    order = None
+    prev_stride = None
+    for stride in ary.strides:
+        if prev_stride:
+            if not order and stride > prev_stride:
+                order = 'F'
+            elif not order and stride < prev_stride:
+                order = 'C'
+            elif ((order == 'F' and stride < prev_stride) or
+                  (order == 'C' and stride > prev_stride)):
+                raise TypeError("Array must have row- or column-major layout")
+        prev_stride = stride
+    return order
+
+def _copy(dst, src):
+    """Copy the contents of src into dst."""
+    if not (isinstance(src, GPUArray) or isinstance(src, np.ndarray)):
```

Comment:
`isinstance(src, (GPUArray, np.ndarray))` is preferable to `or`-d together single `isinstance()`, since `isinstance` is kind of a slow operation.

That said, the host side should not really be limited to numpy arrays--in principle anything with a buffer interface is OK.